lib/postgrex.ex

```diff
@@ -269,7 +272,8 @@ defmodule Postgrex do
   Runs an (extended) prepared query and returns the result or raises
   `Postgrex.Error` if there was an error. See `execute/4`.
   """
-  @spec execute!(conn, Postgrex.Query.t, list, Keyword.t) :: Postgrex.Result.t
+  @spec execute!(conn, Postgrex.Query.t, list, Keyword.t) ::
+    Postgrex.Result.t | {Postgrex.Query.t, Postgrex.Result.t}
```

Comment:
I'm a little worried about this interface change. It can go unnoticed for a long time and lead to subtle and hard to detect bugs.

Reply:
Do you want the query to only be returned for non-bang version?

Reply:
I'm not sure, but I would say so. With the non-bang version, the user most probably already pattern matches on the result, so the change will be noticed quickly. With `execute!` it's likely the result is not pattern matched and the error could appear somewhere completely unrelated to the call.

Reply:
The trouble here is that if someone is using `execute!` (or preparing queries in anyway that last for longer than a single transaction) then they can't handle type changes unless they abuse internals like Ecto does. Of course Ecto can only really handle this because development is synced by `elixir-ecto` ;). Returning a query is an alternative to raising an error here. If we ignore the query then all executes take 2 roundtrips from this point forward.

Reply:
Shall we add an `execute` option to enable this feature such as `redescribe: boolean` that defaults to false.

Reply:
Alternatively, we could have the `execute!` call start to always return a tuple. This should also protect from those "subtle bugs". I'm not sure, though, what we would return as the second element in that case.

Reply:
We are able to use the `ref` value to quickly compare difference between query structs so a fast comparison can be done. It seems preferable than sometimes returning the query as testing that is harder. However it is breaking the API.

We could also ignore the query struct return, like we do for `query`, and continue to abuse internals in Ecto. Then we delay this change until we have a complete solution: https://github.com/elixir-ecto/db_connection/issues/102.

Reply:
We'll ignore the query struct for now but can change this later.